core/src/main/scala/kafka/log/Log.scala

```diff
@@ -792,6 +798,34 @@ class Log(@volatile var dir: File,
     deleteOldSegments(shouldDelete)
   }
 
+  private def updateAppendInfoForDuplicateEntry(appendInfo: LogAppendInfo) : LogAppendInfo = {
+    val segmentIterator = segments.descendingMap().values().iterator()
+    while (segmentIterator.hasNext) {
+      val segment = segmentIterator.next()
+      val firstSequence = getFirstSequence(segment)
+      if (0 <= firstSequence && firstSequence <= appendInfo.firstSequence) {
+        for (entry <- segment.log.entries.asScala) {
+          if (entry.firstSequence.equals(appendInfo.firstSequence)) {
+            appendInfo.firstOffset = entry.baseOffset
+            appendInfo.lastOffset = entry.lastOffset
+            if (entry.timestampType().equals(TimestampType.LOG_APPEND_TIME))
+              appendInfo.logAppendTime = entry.maxTimestamp
+            return appendInfo
+          }
+        }
+      }
+    }
+    null
+  }
+
+  private def getFirstSequence(segment: LogSegment) : Int = {
```

Comment:
This method needs to take the PID. Also, we could return `Option` and use `None` instead of -1.